src/cascadia/TerminalSettingsModel/CascadiaSettingsSerialization.cpp

```diff
@@ -293,7 +293,17 @@ void SettingsLoader::FinalizeLayering()
     for (const auto& profile : userSettings.profiles)
     {
         profile->AddMostImportantParent(userSettings.baseLayerProfile);
+
+        // This completes the parenting process that was started in _addUserProfileParent().
         profile->_FinalizeInheritance();
+        if (profile->Origin() == OriginTag::None)
+        {
+            // If you add more fields here, make sure to do the same in
+            // implementation::CreateChild().
+            profile->Origin(OriginTag::User);

```

Comment:
So, wait, can you walk me through this?

We load a generated profile, and then a fragment. We have one origin::generated, one origin::fragment.
Then we make a new child profile of those, right? _Does it have `OriginTag::None`_?

Is it _correct_ to copy the name up to the user-owned leaf node?

What caused this bug originally? The user-owned leaf node wasn't being assigned a name, was it? Or, was it being assigned a name too early?

Reply:
> We load a generated profile, and then a fragment. We have one origin::generated, one origin::fragment.
Then we make a new child profile of those, right? Does it have `OriginTag::None`?

Yes, temporarily at least.

> Is it _correct_ to copy the name up to the user-owned leaf node?

You mean the profile that gets written into settings.json? Yeah, this is equivalent to the previous code and the old code before the refactorization.

> What caused this bug originally? The user-owned leaf node wasn't being assigned a name, was it? Or, was it being assigned a name too early?

Assigning a name too early.

----

The rather lengthy comment in `_addUserProfileParent` explains this somewhat... hopefully.
This code is executed whenever we encounter a new user profile - one that truly doesn't exist yet in the user's settings.json. In these cases we'll make a new, empty child profile and insert it as a user profile. The `profile` given as a parameter to `_addUserProfileParent` is the first parent of our new user profile (hence the function name).

However unlike the old code, we shouldn't simply call `CreateChild()` here. Under the assumption that the given profile has a finalized inheritance, `CreateChild()` copies the name and hidden fields over into the child. If we add more parents to our user profile however (fragments basically), then the value for name and hidden might change still. Due to this we'll have to defer copying any attributes for as long as possible and can't use `CreateChild()` which would do it early.

The way we detect these new user profiles is that they don't have an origin yet.

Reply:
Scary. Okay, I understand.

Reply:
I thought about adding another attribute to Profile, saying that it's not finished initializing yet, but I felt like that's just as bad of a hack.